README.md
# influxdb-relay

[![License][license-img]][license-href]

1. [Overview](#overview)
2. [Description](#description)
3. [Requirements](#requirements)
4. [Setup](#setup)
5. [Usage](#usage)
6. [Limitations](#limitations)
7. [Development](#development)
8. [Miscellaneous](#miscellaneous)

## Overview

Maintained fork of [influxdb-relay][overview-href] originally developed by
InfluxData. Replicate InfluxDB data for high availability.

## Description

This project adds a basic high availability layer to InfluxDB. With the right
architecture and disaster recovery processes, this achieves a highly available
setup.

## Tested on

- [Go](https://golang.org/doc/install) 1.7.4 to 1.11.2
- [InfluxDB](https://docs.influxdata.com/influxdb/v1.6/introduction/installation/) 1.6.4 to 1.7

Other versions will probably work but are untested.

## Setup

### Go

Download the daemon into your `$GOPATH` and install it in `/sur/sbin`.

```sh
go get -u github.com/vente-privee/influxdb-relay
cp ${GOPATH}/bin/influxdb-relay /usr/bin/influxdb-relay
chmod 755 /usr/bin/influxdb-relay
```

Create the configuration file in `/etc/influxdb-relay`.

```sh
mkdir -p /etc/influxdb-relay
cp ${GOPATH}/src/github.com/vente-privee/influxdb-relay/examples/sample.conf \
   /etc/influxdb-relay/influxdb-relay.conf
```

### Docker

Build your own image.

```sh
git clone git@github.com:vente-privee/influxdb-relay
cd influxdb-relay
docker build \
       --file Dockerfile \
       --rm \
       --tag influxdb-relay:latest \
       .
docker run \
       --volume /path/to/influxdb-relay.conf:/etc/influxdb-relay/influxdb-relay.conf
       --rm
       influxdb-relay:latest
```

or

Docker pull our image.

```sh
docker pull vptech/influxdb-relay:latest
docker run \
       --volume /path/to/influxdb-relay.conf:/etc/influxdb-relay/influxdb-relay.conf
       --rm
       vptech/influxdb-relay:latest
```

## Usage

You can find more documentation in [docs](docs) folder.

* [Architecture](docs/architecture.md)
* [Buffering](docs/buffering.md)
* [Caveats](docs/caveats.md)
* [Recovery](docs/recovery.md)
* [Sharding](docs/sharding.md)

You can find some configurations in [examples](examples) folder.

* [sample.conf](examples/sample.conf)
* [sample_buffered.conf](examples/sample_buffered.conf)
* [kapacitor.conf](examples/kapacitor.conf)

### Configuration

```toml
[[http]]
# Name of the HTTP server, used for display purposes only.
name = "example-http"

# TCP address to bind to, for HTTP server.
bind-addr = "127.0.0.1:9096"

# Ping response code, default is 204
default-ping-response = 200

# Enable HTTPS requests.
ssl-combined-pem = "/path/to/influxdb-relay.pem"

# Array of InfluxDB instances to use as backends for Relay.
output = [
    # name: name of the backend, used for display purposes only.
    # location: full URL of the /write endpoint of the backend
    # admin: Partial URL of the backend, without any path
    # timeout: Go-parseable time duration. Fail writes if incomplete in this time.
    # skip-tls-verification: skip verification for HTTPS location. WARNING: it's insecure. Don't use in production.
    # type: type of input source. OPTIONAL: see below for more information.

    # InfluxDB
    { name="local-influxdb01", location="http://127.0.0.1:8086/write", timeout="10s", type="influxdb" },
    { name="local-influxdb02", location="http://127.0.0.1:7086/write", timeout="10s", type="influxdb" },

    # Prometheus
    { name="local-influxdb03", location="http://127.0.0.1:6086/api/v1/prom/write", timeout="10s", type="prometheus" },
    { name="local-influxdb05", location="http://127.0.0.1:5086/api/v1/prom/write", timeout="10s", type="prometheus" },
]

[[udp]]
# Name of the UDP server, used for display purposes only.
name = "example-udp"

# UDP address to bind to.
bind-addr = "127.0.0.1:9096"

# Socket buffer size for incoming connections.
read-buffer = 0 # default

# Precision to use for timestamps
precision = "n" # Can be n, u, ms, s, m, h

# Array of InfluxDB instances to use as backends for Relay.
output = [
    # name: name of the backend, used for display purposes only.
    # location: host and port of backend.
    # mtu: maximum output payload size
    { name="local1", location="127.0.0.1:8089", mtu=512 },
    { name="local2", location="127.0.0.1:7089", mtu=1024 },
]
```

InfluxDB Relay is able to forward from a variety of input sources, including:

* `influxdb`
* `prometheus`

The `type` parameter in the configuration file defaults to `influxdb`.

### Administrative tasks

#### /admin endpoint

Whereas data manipulation relies on the `/write` endpoint, some other features
such as database or user management are based on the `/query` endpoint. As
InfluxDB Relay does not send back a response body to the client(s), we are not
able to forward all of the features this endpoint provides. Still, we decided
to expose it through the `/admin` route.

In order to use it, one shall define the optional `admin` parameter of the
target backend(s) in the configuration file, as follows:

```toml
[[http]]
# Name of the HTTP server, used for display purposes only.
name = "expose-database-creation"

# TCP address to bind to, for HTTP server.
bind-addr = "127.0.0.1:9096"

# Array of InfluxDB instances to use as backends for Relay.
output = [
    # InfluxDB
    { name="local-influxdb01", location="http://127.0.0.1:8086/write", admin="http://127.0.0.1:8086", type="influxdb" },
    { name="local-influxdb02", location="http://127.0.0.1:7086/write", admin="http://127.0.0.1:7086", type="influxdb" },
]
```

It is now possible to query the `/admin` endpoint. Its usage is the same as the
standard `/query` Influx DB enpoint:

```
curl -X POST "http://127.0.0.1:9096/admin" --data-urlencode 'q=CREATE DATABASE some_database'
```

Errors will be logged just like regular `/write` queries. The HTTP response
bodies will not be forwarded back to the clients.

#### /health endpoint

This endpoint uses the same configuration as the `/admin` endpoint, and provides a quick way to check the state of all the backends.
It will return a JSON object detailing the status of the backends like this : 

```json
{
  "status": "problem",
  "problem": {
    "local-influxdb01": "KO. Get http://influxdb/ping: dial tcp: lookup influxdb on 0.0.0.0:8086: no such host"
  },
  "healthy": {
    "local-influxdb02": "OK. Time taken 3ms"
  }
}
```

If the relay encounters an error while checking a backend, this backend will be reported with the associated error in the `problems` object.
The backends wich the relay was able to communicate with will be reported in the healthy object.

The status field is a summary of the general state of the backends, the defined states are as follows:
* `healthy`: no errors were encountered
* `problem`: some backends, but no all of them, returned errors
* `critical`: every backend returned an error

## Limitations

So far, this is compatible with Debian, RedHat, and other derivatives.

Moreover, the `type` parameter is not supported when it comes to UDP forwarding.

## Development

Please read carefully [CONTRIBUTING.md][contribute-href] before making a merge
request.

Clone repository into your `$GOPATH`.

```sh
mkdir -p ${GOPATH}/src/github.com/vente-privee
cd ${GOPATH}/src/github.com/vente-privee
git clone git@github.com:vente-privee/influxdb-relay
```

Enter the directory and build the daemon.

```sh
cd ${GOPATH}/src/github.com/vente-privee/influxdb-relay
go build -a -ldflags '-extldflags "-static"' -o influxdb-relay
```

## Miscellaneous

```
    ╚⊙ ⊙╝
  ╚═(███)═╝
 ╚═(███)═╝
╚═(███)═╝
 ╚═(███)═╝
  ╚═(███)═╝
   ╚═(███)═╝
```

[license-img]: https://img.shields.io/badge/license-MIT-blue.svg
[license-href]: LICENSE
[overview-href]: https://github.com/influxdata/influxdb-relay
[contribute-href]: CONTRIBUTING.md


CHANGELOG.md
Jun 28 2018 Antoine MILLET <amillet@vente-privee.com>
	From https://github.com/influxdata/influxdb-relay

Jun 29 2018 Alexandre BESLIC <abeslic@abronan.com>
	* Switch to dep, make the project buildable after go get

Jun 29 2018 Dejan FILIPOVIC <dfilipovic@vente-privee.com>
	* Add /status route

Jun 30 2018 Antoine MILLET <amillet@vente-privee.com>
	* New README structure
	* Add basic tests with golint & pylint
	* Add CHANGELOG
	* Add CONTRIBUTING guide
	* Merge https://github.com/influxdata/influxdb-relay/pull/65
	* Merge https://github.com/influxdata/influxdb-relay/pull/52
	* Merge https://github.com/influxdata/influxdb-relay/pull/59
	* Merge https://github.com/influxdata/influxdb-relay/pull/43
	* Merge https://github.com/influxdata/influxdb-relay/pull/57

Nov 15 2018 Maxime CORBIN <mcorbin@vente-privee.com>
    * Add Prometheus input support
    * Add `/admin` route to administrate underlying databases
    * Add code coverage / unit tests
    * Code refactor
    * Improve buffering feature avoiding connexions hanging
    * Improve `/ping` route
    * Improve logging
    * Add `-version` option
    
Dec 13 2018 Maxime CORBIN <mcorbin@vente-privee.com>
    * Added `/admin` endpoint that can be used to create or remove databases
    * Added `/health` endpoint that can be used to monitor the health status of every backend
    * Fixed some performance bugs & added a few more logs

relay/http_middlewares_test.go
package relay

import (
	"bytes"
	"compress/gzip"
	"github.com/stretchr/testify/assert"
	"github.com/vente-privee/influxdb-relay/config"
	"io/ioutil"
	"log"
	"net/http"
	"testing"
)

// Body is an empty body which implements io.Reader in order to create valid http.Requests
type Body struct {
	buf *bytes.Buffer
}

func (b Body) Read(p []byte) (int, error)   { return b.buf.Read(p) }
func (b *Body) Write(p []byte) (int, error) { return b.buf.Write(p) }
func (b Body) Close() error                 { return nil }

// End is a placeholder http Handler used to test middlewares
func (h *HTTP) End(w http.ResponseWriter, r *http.Request) {
	return
}

// EndTag is a placeholder http Handler used to test middlewares, which changes a variable
func (h *HTTP) EndTag(w http.ResponseWriter, r *http.Request) {
	wasInEnd = true
	return
}

func (h *HTTP) tagMiddleware(next relayHandlerFunc) relayHandlerFunc {
	return relayHandlerFunc(func(h *HTTP, w http.ResponseWriter, r *http.Request) {
		wasInMiddleware = true
		next(h, w, r)
	})
}

// ResponseWriter is a placeholder http.ResponseWriter to give to middlewares
type ResponseWriter struct {
	writeBuf *bytes.Buffer
	code     int
	header   http.Header
}

func (r *ResponseWriter) Header() http.Header         { return r.header }
func (r *ResponseWriter) Write(p []byte) (int, error) { return r.writeBuf.Write(p) }
func (r *ResponseWriter) WriteHeader(statusCode int)  { r.code = statusCode }

type Logger struct {
	buffer *bytes.Buffer
}

func (l Logger) Write(p []byte) (n int, err error) { return l.buffer.Write(p) }

var (
	emptyBody     = Body{}
	wrongGzipBody = Body{}
	gzipBody      = Body{}

	wasInMiddleware = false
	wasInEnd        = false

	logger = Logger{buffer: bytes.NewBuffer([]byte{})}

	emptyWriter = &ResponseWriter{
		writeBuf: &bytes.Buffer{},
		header:   http.Header{},
	}

	errorDbWriter = &ResponseWriter{
		writeBuf: bytes.NewBuffer([]byte("\"missing parameter: db\"")),
		header: http.Header{
			"Content-Length": []string{"23"},
			"Content-Type":   []string{"application/json"},
		},
		code: 400,
	}

	errorGzipWriter = &ResponseWriter{
		writeBuf: bytes.NewBuffer([]byte("\"unable to decode gzip body\"")),
		header: http.Header{
			"Content-Length": []string{"28"},
			"Content-Type":   []string{"application/json"},
		},
		code: 400,
	}
)

var (
	// w is a ResponseWriter that will be given to the middlewares
	w = &ResponseWriter{
		writeBuf: &bytes.Buffer{},
		header:   http.Header{},
	}
)

func init() {
	wrongGzipBody.buf = &bytes.Buffer{}
	_, err := wrongGzipBody.Write([]byte("WRONG"))
	if err != nil {
		log.Fatal(err)
	}

	gzipBody.buf = &bytes.Buffer{}
	gzipWriter := gzip.NewWriter(gzipBody.buf)
	_, err = gzipWriter.Write([]byte("GOOOOD"))
	if err != nil {
		log.Fatal(err)
	}
	if err = gzipWriter.Close(); err != nil {
		log.Fatal(err)
	}
}

func resetWriter() {
	w.writeBuf = &bytes.Buffer{}
	w.code = 0
	w.header = http.Header{}
}

func WriterTest(t *testing.T, expected *ResponseWriter, actual *ResponseWriter) {
	assert.Equal(t, expected.code, actual.code)
	expectedBuf, _ := ioutil.ReadAll(expected.writeBuf)
	actualBuf, _ := ioutil.ReadAll(actual.writeBuf)
	assert.Equal(t, string(expectedBuf), string(actualBuf))
	assert.Equal(t, expected.header, actual.header)
}

func createHTTP(t *testing.T, cfg config.HTTPConfig, verbose bool) *HTTP {
	tmp, err := NewHTTP(cfg, verbose)
	if err != nil {
		t.Fatal(err)
	}
	return tmp.(*HTTP)
}

func TestLogMiddleware(t *testing.T) {
	defer resetWriter()
	h := createHTTP(t, config.HTTPConfig{}, true)

	h.logger = log.New(logger, "", 0)
	handler := h.logMiddleWare((*HTTP).End)
	r, err := http.NewRequest("", "influxdb", emptyBody)
	if err != nil {
		t.Fatal(err)
	}
	handler(h, w, r)
	buf, _ := ioutil.ReadAll(logger.buffer)
	assert.Equal(t, string(buf), "Got request on: influxdb\n")
}

func TestLogMiddlewareNoLog(t *testing.T) {
	defer resetWriter()
	h := createHTTP(t, config.HTTPConfig{}, false)

	h.logger = log.New(logger, "", 0)
	handler := h.logMiddleWare((*HTTP).End)
	r, err := http.NewRequest("", "influxdb", emptyBody)
	if err != nil {
		t.Fatal(err)
	}
	handler(h, w, r)
	buf, _ := ioutil.ReadAll(logger.buffer)
	assert.Equal(t, string(buf), "")
}

func TestQueryMiddleware(t *testing.T) {
	defer resetWriter()
	h := createHTTP(t, config.HTTPConfig{}, false)

	handler := h.queryMiddleWare((*HTTP).End)
	r, err := http.NewRequest("", "influxdb", emptyBody)
	if err != nil {
		t.Fatal(err)
	}
	url, _ := r.URL.Parse("http://influxdb:8086/write?db=mydb")
	r.URL = url
	handler(h, w, r)
	assert.Equal(t, emptyWriter, w)
}

func TestQueryMiddlewareNoDB(t *testing.T) {
	defer resetWriter()
	h := createHTTP(t, config.HTTPConfig{}, false)

	handler := h.queryMiddleWare((*HTTP).End)
	r, err := http.NewRequest("", "influxdb", emptyBody)
	if err != nil {
		t.Fatal(err)
	}
	url, _ := r.URL.Parse("http://influxdb:8086/write")
	r.URL = url
	handler(h, w, r)
	WriterTest(t, errorDbWriter, w)
}

func TestQueryMiddlewareRp(t *testing.T) {
	defer resetWriter()
	h := createHTTP(t, config.HTTPConfig{DefaultRetentionPolicy: "patate"}, false)

	handler := h.queryMiddleWare((*HTTP).End)
	r, err := http.NewRequest("", "influxdb", emptyBody)
	if err != nil {
		t.Fatal(err)
	}
	url, _ := r.URL.Parse("http://influxdb:8086/write?db=mydb")
	r.URL = url
	handler(h, w, r)
	assert.Equal(t, emptyWriter, w)
	queryParams := r.URL.Query()
	assert.Equal(t, "patate", queryParams.Get("rp"))
}

func TestBodyMiddleWareNoGzip(t *testing.T) {
	defer resetWriter()
	h := createHTTP(t, config.HTTPConfig{}, false)

	handler := h.bodyMiddleWare((*HTTP).End)
	r, err := http.NewRequest("", "influxdb", emptyBody)
	if err != nil {
		t.Fatal(err)
	}
	handler(h, w, r)
	assert.Equal(t, emptyWriter, w)
}

func TestBodyMiddleWareErrorGzip(t *testing.T) {
	defer resetWriter()
	h := createHTTP(t, config.HTTPConfig{}, false)

	handler := h.bodyMiddleWare((*HTTP).End)
	r, err := http.NewRequest("", "influxdb", wrongGzipBody)
	if err != nil {
		t.Fatal(err)
	}
	r.Header.Set("Content-Encoding", "gzip")
	handler(h, w, r)
	WriterTest(t, errorGzipWriter, w)
}

func TestBodyMiddleWareGzip(t *testing.T) {
	defer resetWriter()
	h := createHTTP(t, config.HTTPConfig{}, false)
	handler := h.bodyMiddleWare((*HTTP).End)
	r, err := http.NewRequest("", "influxdb", gzipBody)
	if err != nil {
		t.Fatal(err)
	}
	r.Header.Set("Content-Encoding", "gzip")
	handler(h, w, r)
	buf, _ := ioutil.ReadAll(r.Body)
	assert.Equal(t, "GOOOOD", string(buf))
}

func TestAllMiddlewares(t *testing.T) {
	h := createHTTP(t, config.HTTPConfig{}, false)
	r, err := http.NewRequest("", "influxdb", gzipBody)
	if err != nil {
		t.Fatal(err)
	}

	middlewares = []relayMiddleware{
		(*HTTP).tagMiddleware,
	}
	chain := allMiddlewares(h, (*HTTP).EndTag)
	chain(h, w, r)
	assert.Equal(t, true, wasInMiddleware)
	assert.Equal(t, true, wasInEnd)
}


main.go
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"os/signal"

	"github.com/vente-privee/influxdb-relay/config"
	"github.com/vente-privee/influxdb-relay/relayservice"
)

const (
	relayVersion = "2.2.0"
)

var (
	usage = func() {
		fmt.Println("Please, see README for more information about InfluxDB Relay...")
		flag.PrintDefaults()
	}

	configFile  = flag.String("config", "", "Configuration file to use")
	verbose     = flag.Bool("v", false, "If set, InfluxDB Relay will log HTTP requests")
	versionFlag = flag.Bool("version", false, "Print current InfluxDB Relay version")
)

func runRelay(cfg config.Config) {
	relay, err := relayservice.New(cfg)
	if err != nil {
		log.Fatal(err)
	}

	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, os.Interrupt)

	go func() {
		<-sigChan
		relay.Stop()
	}()

	log.Println("Starting relays...")
	relay.Run()
}

func main() {
	flag.Usage = usage
	flag.Parse()

	if *versionFlag {
		fmt.Println("influxdb-relay version " + relayVersion)
		return
	}

	// Configuration file is mandatory
	if *configFile == "" {
		fmt.Fprintln(os.Stderr, "Missing configuration file")
		flag.PrintDefaults()
		os.Exit(1)
	}

	// And it has to be loaded in order to continue
	cfg, err := config.LoadConfigFile(*configFile)
	if err != nil {
		log.Fatal(err.Error())
	}

	cfg.Verbose = *verbose
	runRelay(cfg)
}


relay/udp.go
package relay

import (
	"bytes"
	"errors"
	"log"
	"net"
	"sync"
	"sync/atomic"
	"time"

	"github.com/influxdata/influxdb/models"

	"github.com/vente-privee/influxdb-relay/config"
)

const (
	defaultMTU = 1024
)

// UDP is a relay for UDP influxdb writes
type UDP struct {
	addr      string
	name      string
	precision string

	closing int64
	l       *net.UDPConn
	c       *net.UDPConn

	backends []*udpBackend
}

// NewUDP -TODO-
func NewUDP(config config.UDPConfig) (Relay, error) {
	u := new(UDP)

	u.name = config.Name
	u.addr = config.Addr
	u.precision = config.Precision

	l, err := net.ListenPacket("udp", u.addr)
	if err != nil {
		return nil, err
	}

	ul, ok := l.(*net.UDPConn)
	if !ok {
		return nil, errors.New("problem listening for UDP")
	}

	if config.ReadBuffer != 0 {
		if err = ul.SetReadBuffer(config.ReadBuffer); err != nil {
			return nil, err
		}
	}

	u.l = ul

	// UDP doesn't really "listen", this just gets us a socket with
	// the local UDP address set to something random
	u.c, err = net.ListenUDP("udp", nil)
	if err != nil {
		return nil, err
	}

	for i := range config.Outputs {
		cfg := &config.Outputs[i]
		if cfg.Name == "" {
			cfg.Name = cfg.Location
		}

		if cfg.MTU == 0 {
			cfg.MTU = defaultMTU
		}

		addr, err := net.ResolveUDPAddr("udp", cfg.Location)
		if err != nil {
			return nil, err
		}

		u.backends = append(u.backends, &udpBackend{u, cfg.Name, addr, cfg.MTU})
	}

	return u, nil
}

// Name -TODO-
func (u *UDP) Name() string {
	if u.name == "" {
		return u.addr
	}
	return u.name
}

// udpPool is used to reuse and auto-size payload buffers, if incoming packets
// are never larger than 2K, then none of the buffers will be larger than that.
// This prevents having to manually tune the UDP buffer size, or having every
// buffer be 64K to hold the maximum possible payload
var udpPool = sync.Pool{New: func() interface{} { return new(bytes.Buffer) }}

func getUDPBuf() *bytes.Buffer {
	return udpPool.Get().(*bytes.Buffer)
}

func putUDPBuf(b *bytes.Buffer) {
	b.Reset()
	udpPool.Put(b)
}

type packet struct {
	timestamp time.Time
	data      *bytes.Buffer
	from      *net.UDPAddr
}

// Run -TODO-
func (u *UDP) Run() error {

	// buffer that can hold the largest possible UDP payload
	var buf [65536]byte

	// arbitrary queue size for now
	queue := make(chan packet, 1024)

	var wg sync.WaitGroup

	go func() {
		for p := range queue {
			u.post(&p)
			wg.Done()
		}
	}()

	log.Printf("Starting UDP relay %q on %v", u.Name(), u.l.LocalAddr())

	for {
		n, remote, err := u.l.ReadFromUDP(buf[:])
		if err != nil {
			if atomic.LoadInt64(&u.closing) == 0 {
				log.Printf("Error reading packet in relay %q from %v: %v", u.name, remote, err)
			} else {
				err = nil
			}
			close(queue)
			wg.Wait()
			return err
		}
		start := time.Now()

		wg.Add(1)

		// copy the data into a buffer and queue it for processing
		b := getUDPBuf()
		b.Grow(n)
		// bytes.Buffer.Write always returns a nil error, and will panic if out of memory
		_, _ = b.Write(buf[:n])
		queue <- packet{start, b, remote}
	}
}

// Stop -TODO-
func (u *UDP) Stop() error {
	atomic.StoreInt64(&u.closing, 1)
	return u.l.Close()
}

func (u *UDP) post(p *packet) {
	points, err := models.ParsePointsWithPrecision(p.data.Bytes(), p.timestamp, u.precision)
	if err != nil {
		log.Printf("Error parsing packet in relay %q from %v: %v", u.Name(), p.from, err)
		putUDPBuf(p.data)
		return
	}

	out := getUDPBuf()
	for _, pt := range points {
		if _, err = out.WriteString(pt.PrecisionString(u.precision)); err != nil {
			break
		}
		if err = out.WriteByte('\n'); err != nil {
			break
		}
	}

	putUDPBuf(p.data)

	if err != nil {
		putUDPBuf(out)
		log.Printf("Error writing points in relay %q: %v", u.Name(), err)
		return
	}

	for _, b := range u.backends {
		if err := b.post(out.Bytes()); err != nil {
			log.Printf("Error writing points in relay %q to backend %q: %v", u.Name(), b.name, err)
		}
	}

	putUDPBuf(out)
}

type udpBackend struct {
	u    *UDP
	name string
	addr *net.UDPAddr
	mtu  int
}

var errPacketTooLarge = errors.New("payload larger than MTU")

func (b *udpBackend) post(data []byte) error {
	var err error
	for len(data) > b.mtu {
		// find the last line that will fit within the MTU
		idx := bytes.LastIndexByte(data[:b.mtu], '\n')
		if idx < 0 {
			// first line is larger than MTU
			return errPacketTooLarge
		}
		_, err = b.u.c.WriteToUDP(data[:idx+1], b.addr)
		if err != nil {
			return err
		}
		data = data[idx+1:]
	}

	_, err = b.u.c.WriteToUDP(data, b.addr)
	return err
}


relay/http.go
package relay

import (
	"bytes"
	"crypto/tls"
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"log"
	"net"
	"net/http"
	"os"
	"strconv"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"github.com/vente-privee/influxdb-relay/config"
)

// HTTP is a relay for HTTP influxdb writes
type HTTP struct {
	addr   string
	name   string
	schema string

	cert string
	rp   string

	pingResponseCode    int
	pingResponseHeaders map[string]string

	closing int64
	l       net.Listener

	backends []*httpBackend

	start  time.Time
	log    bool
	logger *log.Logger
}

type relayHandlerFunc func(h *HTTP, w http.ResponseWriter, r *http.Request, start time.Time)
type relayMiddleware func(h *HTTP, handlerFunc relayHandlerFunc) relayHandlerFunc

// Default HTTP settings and a few constants
const (
	DefaultHTTPPingResponse = http.StatusNoContent
	DefaultHTTPTimeout      = 10 * time.Second
	DefaultMaxDelayInterval = 10 * time.Second
	DefaultBatchSizeKB      = 512

	KB = 1024
	MB = 1024 * KB
)

var (
	handlers = map[string]relayHandlerFunc{
		"/write":             (*HTTP).handleStandard,
		"/api/v1/prom/write": (*HTTP).handleProm,
		"/ping":              (*HTTP).handlePing,
		"/status":            (*HTTP).handleStatus,
		"/admin":             (*HTTP).handleAdmin,
		"/health":            (*HTTP).handleHealth,
	}

	middlewares = []relayMiddleware{
		(*HTTP).bodyMiddleWare,
		(*HTTP).queryMiddleWare,
		(*HTTP).logMiddleWare,
	}
)

// NewHTTP creates a new HTTP relay
// This relay will most likely be tied to a RelayService
// and manage a set of HTTPBackends
func NewHTTP(cfg config.HTTPConfig, verbose bool) (Relay, error) {
	h := new(HTTP)

	h.addr = cfg.Addr
	h.name = cfg.Name
	h.log = verbose
	if verbose {
		h.logger = log.New(os.Stdout, "", 0)
	}

	h.pingResponseCode = DefaultHTTPPingResponse
	if cfg.DefaultPingResponse != 0 {
		h.pingResponseCode = cfg.DefaultPingResponse
	}

	h.pingResponseHeaders = make(map[string]string)
	h.pingResponseHeaders["X-InfluxDB-Version"] = "relay"
	if h.pingResponseCode != http.StatusNoContent {
		h.pingResponseHeaders["Content-Length"] = "0"
	}

	h.cert = cfg.SSLCombinedPem
	h.rp = cfg.DefaultRetentionPolicy

	// If a cert is specified, this means the user
	// wants to do HTTPS
	h.schema = "http"
	if h.cert != "" {
		h.schema = "https"
	}

	// For each output specified in the config, we are going to create a backend
	for i := range cfg.Outputs {
		backend, err := newHTTPBackend(&cfg.Outputs[i])
		if err != nil {
			return nil, err
		}

		h.backends = append(h.backends, backend)
	}

	return h, nil
}

// Name is the name of the HTTP relay
// a default name might be generated if it is
// not specified in the configuration file
func (h *HTTP) Name() string {
	if h.name == "" {
		return fmt.Sprintf("%s://%s", h.schema, h.addr)
	}

	return h.name
}

// Run actually launch the HTTP endpoint
func (h *HTTP) Run() error {
	var cert tls.Certificate
	l, err := net.Listen("tcp", h.addr)
	if err != nil {
		return err
	}

	// support HTTPS
	if h.cert != "" {
		cert, err = tls.LoadX509KeyPair(h.cert, h.cert)
		if err != nil {
			return err
		}

		l = tls.NewListener(l, &tls.Config{
			Certificates: []tls.Certificate{cert},
		})
	}

	h.l = l

	log.Printf("Starting %s relay %q on %v", strings.ToUpper(h.schema), h.Name(), h.addr)

	err = http.Serve(l, h)
	if atomic.LoadInt64(&h.closing) != 0 {
		return nil
	}
	return err
}

// Stop actually stops the HTTP endpoint
func (h *HTTP) Stop() error {
	atomic.StoreInt64(&h.closing, 1)
	return h.l.Close()
}

// ServeHTTP is the function that handles the different route
// The response is a JSON object describing the state of the operation
func (h *HTTP) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	//h.start = time.Now()

	if fun, ok := handlers[r.URL.Path]; ok {
		allMiddlewares(h, fun)(h, w, r, time.Now())
	} else {
		jsonResponse(w, response{http.StatusNotFound, http.StatusText(http.StatusNotFound)})
		return
	}
}

type responseData struct {
	ContentType     string
	ContentEncoding string
	StatusCode      int
	Body            []byte
}

func (rd *responseData) Write(w http.ResponseWriter) {
	if rd.ContentType != "" {
		w.Header().Set("Content-Type", rd.ContentType)
	}

	if rd.ContentEncoding != "" {
		w.Header().Set("Content-Encoding", rd.ContentEncoding)
	}

	w.Header().Set("Content-Length", strconv.Itoa(len(rd.Body)))
	w.WriteHeader(rd.StatusCode)
	w.Write(rd.Body)
}

func jsonResponse(w http.ResponseWriter, r response) {
	data, err := json.Marshal(r.body)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("Content-Length", fmt.Sprint(len(data)))
	w.WriteHeader(r.code)

	_, _ = w.Write(data)
}

type poster interface {
	post([]byte, string, string) (*responseData, error)
	getStats() map[string]string
}

type simplePoster struct {
	client   *http.Client
	location string
}

func newSimplePoster(location string, timeout time.Duration, skipTLSVerification bool) *simplePoster {
	// Configure custom transport for http.Client
	// Used for support skip-tls-verification option
	transport := &http.Transport{
		TLSClientConfig: &tls.Config{
			InsecureSkipVerify: skipTLSVerification,
		},
	}

	return &simplePoster{
		client: &http.Client{
			Timeout:   timeout,
			Transport: transport,
		},
		location: location,
	}
}

func (s *simplePoster) getStats() map[string]string {
	v := make(map[string]string)
	v["location"] = s.location
	return v
}

func (s *simplePoster) post(buf []byte, query string, auth string) (*responseData, error) {
	req, err := http.NewRequest("POST", s.location, bytes.NewReader(buf))
	if err != nil {
		return nil, err
	}

	req.URL.RawQuery = query
	req.Header.Set("Content-Type", "text/plain")
	req.Header.Set("Content-Length", strconv.Itoa(len(buf)))
	if auth != "" {
		req.Header.Set("Authorization", auth)
	}

	resp, err := s.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	data, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	return &responseData{
		ContentType:     resp.Header.Get("Content-Type"),
		ContentEncoding: resp.Header.Get("Content-Encoding"),
		StatusCode:      resp.StatusCode,
		Body:            data,
	}, nil
}

type httpBackend struct {
	poster
	name      string
	inputType config.Input
	admin     string
}

func newHTTPBackend(cfg *config.HTTPOutputConfig) (*httpBackend, error) {
	// Get default name
	if cfg.Name == "" {
		cfg.Name = cfg.Location
	}

	timeout := DefaultHTTPTimeout
	if cfg.Timeout != "" {
		t, err := time.ParseDuration(cfg.Timeout)
		if err != nil {
			return nil, fmt.Errorf("error parsing HTTP timeout '%v'", err)
		}
		timeout = t
	}

	var p poster = newSimplePoster(cfg.Location, timeout, cfg.SkipTLSVerification)

	// If configured, create a retryBuffer per backend.
	// This way we serialize retries against each backend.
	if cfg.BufferSizeMB > 0 {
		max := DefaultMaxDelayInterval
		if cfg.MaxDelayInterval != "" {
			m, err := time.ParseDuration(cfg.MaxDelayInterval)
			if err != nil {
				return nil, fmt.Errorf("error parsing max retry time %v", err)
			}
			max = m
		}

		batch := DefaultBatchSizeKB * KB
		if cfg.MaxBatchKB > 0 {
			batch = cfg.MaxBatchKB * KB
		}

		p = newRetryBuffer(cfg.BufferSizeMB*MB, batch, max, p)
	}

	return &httpBackend{
		poster:    p,
		name:      cfg.Name,
		inputType: cfg.InputType,
		admin:     cfg.Admin,
	}, nil
}

// ErrBufferFull error indicates that retry buffer is full
var ErrBufferFull = errors.New("retry buffer full")

var bufPool = sync.Pool{New: func() interface{} { return new(bytes.Buffer) }}

func getBuf() *bytes.Buffer {
	if bb, ok := bufPool.Get().(*bytes.Buffer); ok {
		return bb
	}
	return new(bytes.Buffer)
}

func putBuf(b *bytes.Buffer) {
	b.Reset()
	bufPool.Put(b)
}
